### PR TITLE
feat: Added Auto Opt, GPU and `jax.Array`

### PR DESCRIPTION
**Pull request by philip-paul-mueller**
This commit addresses the following issues:
- It adds an implementation for auto optimizer (https://github.com/GridTools/jace/issues/24).
    The current implementation is not likely to stay, since it essentially uses DaCe's version, which is known to have problems with JaCe's SDFGs.
- It allows to run stuff on GPU (https://github.com/GridTools/jace/issues/25).
    While it is possible it is still needed that the user explicitly specify it, JAX does an auto detection.
- Instead of returning NumPy arrays JaCe now returns `jax.Array` objects (https://github.com/GridTools/jace/issues/22).
    This goes in tandem with a reworking of the type annotation, which was wrong before (and can not be correctly made).

### CODE REVIEW

File: src/jace/api.py [hunk 1]
```diff
@@ -22,44 +22,47 @@
...
-    Note:
-        Currently there are no known options, but essentially it is a subset of some
-        of the options that are supported by `jax.jit` together with some additional
-        JaCe specific ones.
+    Not all arguments that are supported by `jax-jit()` are also supported by
+    `jace.jit`. Furthermore, some additional ones might be supported.
+    The following arguments are supported:
+    - `backend`: For which platform DaCe should generate code for. It is a string,
+        where the following values are supported: `'cpu'` or `'gpu'`.
+        DaCe's `DeviceType` enum or FPGA are not supported.
```
Comment (egparedes) [inline src/jace/api.py:36]:
```suggestion
    Not all arguments that are supported by `jax.jit()` are also supported by
    `jace.jit()` and some additional ones might be added.
    
    Args:
        backend: Target platform for which DaCe should generate code.
```

File: src/jace/api.py [hunk 2]
```diff
@@ -22,44 +22,47 @@
...
     """
 
+    backend: str
```
Comment (egparedes) [inline src/jace/api.py:39]:
```suggestion
    backend: Literal['cpu', 'gpu']
```

File: src/jace/api.py [hunk 3]
```diff
@@ -72,18 +75,19 @@ def jit(
         fun: Function to wrap.
         primitive_translators: Use these primitive translators for the lowering to SDFG.
             If not specified the translators in the global registry are used.
-        kwargs: Jit arguments.
+        kwargs: Jit arguments, see `JITOptions` for more.
```
Comment (egparedes) [inline src/jace/api.py:78]:
```suggestion
        kwargs: jit arguments, see `JITOptions` for more.
```

File: src/jace/api.py [hunk 4]
```diff
@@ -72,18 +75,19 @@ def jit(
...
 
     Note:
         This function is the only valid way to obtain a JaCe computation.
     """
-    if kwargs:
+    not_supported_jit_keys = kwargs.keys() - {"backend"}
```
Comment (egparedes) [inline src/jace/api.py:83]: Wouldn't it be better to use the information in the TypedDict?
```suggestion
    not_supported_jit_keys = kwargs.keys() - JITOptions.__annotations__.keys()
```

File: src/jace/api.py [hunk 5]
```diff
@@ -72,18 +75,19 @@ def jit(
...
+    if not_supported_jit_keys:
         # TODO(phimuell): Add proper name verification and exception type.
         raise NotImplementedError(
-            f"The following arguments to 'jace.jit' are not yet supported: {', '.join(kwargs)}."
+            f"The following arguments to 'jace.jit' are not yet supported: {', '.join(not_supported_jit_keys)}."
```
Comment (egparedes) [inline src/jace/api.py:87]: I would also raise a `RuntimeError` or a `ValueError`
```suggestion
            f"The following arguments to 'jace.jit' are not supported: {', '.join(not_supported_jit_keys)}."
```

File: src/jace/optimization.py [hunk 6]
```diff
@@ -62,22 +74,46 @@ def jace_optimize(tsdfg: tjsdfg.TranslatedJaxprSDFG, **kwargs: Unpack[CompilerOp
...
-        auto_optimize: Run the auto optimization pipeline (currently does nothing)
+        auto_optimize: Run the auto optimization pipeline.
         persistent_transients: Set the allocation lifetime of (non register) transients
             in the SDFG to `AllocationLifetime.Persistent`, i.e. keep them allocated
             between different invocations.
-    """
-    # TODO(phimuell): Implement the functionality.
-    # Currently this function exists primarily for the sake of existing.
+        validate: Perform validation at the end.
+        validate_all: Perform extensive validation.
```
Comment (egparedes) [inline src/jace/optimization.py:84]: Ambiguous documentation, especially for `validate_all`

File: src/jace/optimization.py [hunk 7]
```diff
@@ -62,22 +74,46 @@ def jace_optimize(tsdfg: tjsdfg.TranslatedJaxprSDFG, **kwargs: Unpack[CompilerOp
...
 
+    Note:
+        Currently DaCe's auto optimization pipeline is used when auto optimize is
+        enabled. However, it might change in the future. Because DaCe's auto
+        optimizer is considered unstable it must be explicitly enabled.
+    """
+    assert device in {dace.DeviceType.CPU, dace.DeviceType.GPU}
     simplify = kwargs.get("simplify", False)
     auto_optimize = kwargs.get("auto_optimize", False)
```
Comment (egparedes) [inline src/jace/optimization.py:93]: Why not?
```suggestion
    simplify = kwargs.get("simplify", DEFAULT_OPTIMIZATIONS["simplify"])
    auto_optimize = kwargs.get("auto_optimize", DEFAULT_OPTIMIZATIONS["auto_optimize"])
```

File: src/jace/optimization.py [hunk 8]
```diff
@@ -62,22 +74,46 @@ def jace_optimize(tsdfg: tjsdfg.TranslatedJaxprSDFG, **kwargs: Unpack[CompilerOp
...
+    validate = kwargs.get("validate", DEFAULT_OPTIMIZATIONS["validate"])
```
Comment (egparedes) [inline src/jace/optimization.py:94]: Why is `persistent_transients` ignored?

File: src/jace/stages.py [hunk 9]
```diff
@@ -166,14 +165,21 @@ def lower(self, *args: _P.args, **kwargs: _P.kwargs) -> JaCeLowered[_R]:
         trans_ctx: translator.TranslationContext = builder.translate_jaxpr(jaxpr)
 
         flat_call_args = jax_tree.tree_leaves((args, kwargs))
+        device = util.parse_backend_jit_option(self._jit_options.get("backend", "cpu"))
```
Comment (egparedes) [inline src/jace/stages.py:168]: Shouldn't the default backend be defined in a single place and then used from where is needed instead of as a hardcoded string in multiple places?

File: src/jace/stages.py [hunk 10]
```diff
@@ -227,20 +234,23 @@ class JaCeLowered(tcache.CachingStage["JaCeCompiled"], Generic[_R]):
...
         tsdfg: tjsdfg.TranslatedJaxprSDFG,
         out_tree: jax_tree.PyTreeDef,
         jaxpr: jax_core.ClosedJaxpr,
+        device: str | dace.DeviceType,
     ) -> None:
         super().__init__()
         self._translated_sdfg = tsdfg
         self._out_tree = out_tree
         self._jaxpr = jaxpr
+        self._device = util.parse_backend_jit_option(device)
```
Comment (egparedes) [inline src/jace/stages.py:250]: This is a matter of style but I'd rather exclude checking the input type inside the parser function: 
```suggestion
        self._device = device if isinstance(str, dace.DeviceType) else util.parse_backend_jit_option(device)
```

File: src/jace/util/jax_helper.py [hunk 11]
```diff
@@ -258,3 +264,42 @@ def get_jax_literal_value(lit: jax_core.Atom) -> bool | float | int | np.generic
     if util.is_scalar(val):
         return val
     raise TypeError(f"Failed to extract value from '{lit}' ('{val}' type: {type(val).__name__}).")
+
+
+def parse_backend_jit_option(
+    backend: str | dace.DeviceType,
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:270]: As mentioned above, I'd rather:
```suggestion
    backend: Literal['cpu', 'gpu'],
```

File: src/jace/util/jax_helper.py [hunk 12]
```diff
@@ -258,3 +264,42 @@ def get_jax_literal_value(lit: jax_core.Atom) -> bool | float | int | np.generic
...
+        return backend
+    match backend:
+        case "cpu" | "CPU":
+            return dace.DeviceType.CPU
+        case "gpu" | "GPU":
+            return dace.DeviceType.GPU
+        case "fpga" | "FPGA":
+            return dace.DeviceType.FPGA
+        case "tpu" | "TPU":
+            raise NotImplementedError("TPU are not supported.")
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:283]: I don't see why this is useful if it is not supported

File: src/jace/util/jax_helper.py [hunk 13]
```diff
@@ -258,3 +264,42 @@ def get_jax_literal_value(lit: jax_core.Atom) -> bool | float | int | np.generic
...
+    By default `copy` is set to `False`, it is the responsibility of the caller
+    to ensure that the underlying buffer is not modified later.
+
+    Args:
+        arr: The array to move into a JAX array.
+        copy: Should a copy be made; defaults to `False`.
+    """
+    if isinstance(arr, jax.Array):
+        return arr
+    # In newer version it is no longer needed to pass a capsule.
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:304]:
```suggestion
```

File: src/jace/util/jax_helper.py [hunk 14]
```diff
@@ -258,3 +264,42 @@ def get_jax_literal_value(lit: jax_core.Atom) -> bool | float | int | np.generic
...
+    return jax_dlpack.from_dlpack(arr, copy=copy)  # type: ignore[attr-defined]  # `from_dlpack` is not found.
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:305]: Why is this type ignore needed if `jax` is included in our lenient mypy settings (and also the function should actually exist)?

File: src/jace/util/jax_helper.py [hunk 15]
```diff
@@ -30,6 +30,12 @@
     import numpy as np
 
 
+try:
+    import cupy as cp  # type: ignore[import-not-found]
+except ImportError:
+    cp = None
+
+
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:38]: Why is this needed?

File: src/jace/tracing.py [hunk 16]
```diff
@@ -77,14 +77,20 @@ def make_jaxpr(
...
-        raise NotImplementedError(
-            f"Not supported tracing options: {', '.join(f'{k}' for k in trace_options)}"
-        )
     # TODO(phimuell): Test if this restriction is needed.
     assert all(param.default is param.empty for param in inspect.signature(fun).parameters.values())
 
+    # NOTE: `jax.make_jaxpr()` to current tracing backend we use, never supported
+    #   all arguments `jax.jit()` supported, which is strange. But in JaCe this
+    #   _will_ not be the case. To make things work, until we have a better working
+    #   backend we have to filter, i.e. clearing the `trace_options`.
```
Comment (egparedes) [inline src/jace/tracing.py:91]: Check and rewrite, I don't understand it.

File: src/jace/tracing.py [hunk 17]
```diff
@@ -96,7 +102,9 @@ def tracer_impl(
         #  it for the tracing.
         with jax.experimental.enable_x64():
             # TODO(phimuell): copy the implementation of the real tracing
-            jaxpr_maker = jax.make_jaxpr(
+            # For some reasons MyPy seems to think that `jax.make_jaxpr()` is the same
+            #  as `jace.make_jaxpr()` so we have to ignore the error.
```
Comment (egparedes) [inline src/jace/tracing.py:106]: Sounds suspicious...

File: src/jace/translator/post_translation.py [hunk 18]
```diff
@@ -133,7 +151,8 @@ def _create_input_state(
 
     The function will create a new set of variables that are exposed as inputs. This
     variables are based on the example input arguments passed through `flat_call_args`.
-    This process will hard code the memory location and strides into the SDFG.
+    This process will hard code the memory location, i.e. if the input is on the GPU,
+    then the new input will be don the GPU as well and strides into the SDFG.
```
Comment (egparedes) [inline src/jace/translator/post_translation.py:155]:
```suggestion
    then the new input will be on the GPU as well and strides into the SDFG.
```

**Review by egparedes — COMMENTED** [review]
Looks good, I have some minor comments. (BTW, Iit is not by any means "a super small PR")

File: src/jace/api.py (hunk 2 above)
Comment (philip-paul-mueller) [reply src/jace/api.py:39]: I do not think that enforcing this through the annotation is the right way.

File: src/jace/api.py (hunk 1 above)
Comment (philip-paul-mueller) [reply src/jace/api.py:36]: I kept some of the old description.

File: src/jace/api.py (hunk 3 above)
Comment (philip-paul-mueller) [reply src/jace/api.py:78]: We should be consistent and write JIT capital or `Jit` as it is the abbreviation of "Just in time" ('Just' is capital because it is the beginning of a sentence).

File: src/jace/api.py (hunk 4 above)
Comment (philip-paul-mueller) [reply src/jace/api.py:83]: Thanks. I was sure that this was possible bud did not know how to do.

File: src/jace/optimization.py (hunk 7 above)
Comment (philip-paul-mueller) [reply src/jace/optimization.py:93]: The documentation says that if not specified it will act as deactivated, thus we must take it from `NO_OPTIMIZATIONS`.
I slightly modified the code and now at the beginning we have `kwargs = {**NO_OPTIMIZIONS, **kwargs}`, which is cleaner.

File: src/jace/optimization.py (hunk 8 above)
Comment (philip-paul-mueller) [reply src/jace/optimization.py:94]: This is a behavior in DaCe I do not like and with this you can disable it, at least in the future this is the plan.

File: src/jace/stages.py (hunk 9 above)
Comment (philip-paul-mueller) [reply src/jace/stages.py:168]: You are right.
I have now introduced a global variable, which contains the string.
I was also able to reduce it to a single point where it is used/needed.

File: src/jace/stages.py (hunk 10 above)
Comment (philip-paul-mueller) [reply src/jace/stages.py:250]: I deleted that line.
However, I think this should be put inside the parse function.
As you have to repeat it everytime you use it.

File: src/jace/util/jax_helper.py (hunk 11 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:270]: See above.

File: src/jace/util/jax_helper.py (hunk 12 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:283]: It is a better error message.
I think the message that JaCe does not support TPU is much better than an error that is indicating that the parsing failed.

File: src/jace/util/jax_helper.py (hunk 14 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:305]: I have no idea.

File: src/jace/util/jax_helper.py (hunk 15 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:38]: In case you want to use JaCe one a CPU only machine as the CI.

File: src/jace/tracing.py (hunk 17 above)
Comment (philip-paul-mueller) [reply src/jace/tracing.py:106]: I know, but I have no idea why.

File: src/jace/tracing.py (hunk 16 above)
Comment (philip-paul-mueller) [reply src/jace/tracing.py:91]: You are right it was written very strangely.

File: src/jace/api.py (hunk 2 above)
Comment (egparedes) [reply src/jace/api.py:41]: Why not? What's the use case for the `Literal` type annotation then? Using `Literal` here also helps in  documenting the supported values in a more prominent way.

File: src/jace/api.py (hunk 3 above)
Comment (egparedes) [reply src/jace/api.py]: I'd rather use `JIT` then but it's up to you...

Comment (egparedes) [reply src/jace/api.py]: I'd rather use `JIT` then but it's up to you...

File: src/jace/tracing.py [hunk 19]
```diff
@@ -77,14 +77,22 @@ def make_jaxpr(
...
-    if trace_options:
-        raise NotImplementedError(
-            f"Not supported tracing options: {', '.join(f'{k}' for k in trace_options)}"
-        )
     # TODO(phimuell): Test if this restriction is needed.
     assert all(param.default is param.empty for param in inspect.signature(fun).parameters.values())
 
+    # NOTE: In the current implementation we are using `jax.make_jaxpr()`. But this
+    #   is a different implementation than `jax.jit()` uses. The main difference
+    #   between the two, seems to be the set of arguments that are supported. In JaCe,
```
Comment (egparedes) [inline src/jace/tracing.py:90]:
```suggestion
    #   between the two seems to be the set of arguments that are supported. In JaCe,
```

File: src/jace/tracing.py [hunk 20]
```diff
@@ -77,14 +77,22 @@ def make_jaxpr(
...
+    #   however, we want to support all arguments that `jace.jit()` does.
+    #   For establishing compatibility we have to clear the arguments to make them
+    #   compatible, with what `jax.make_jaxpr()` and `jace.jit()` supports.
```
Comment (egparedes) [inline src/jace/tracing.py:93]:
```suggestion
    #   compatible with what `jax.make_jaxpr()` and `jace.jit()` supports.
```

File: src/jace/util/jax_helper.py (hunk 15 above)
Comment (egparedes) [reply src/jace/util/jax_helper.py:38]: I meant, why is cupy needed here? I can't see where it is used in this module.

File: src/jace/util/jax_helper.py (hunk 12 above)
Comment (egparedes) [reply src/jace/util/jax_helper.py:283]: This function should accept only strings or it should be renamed, because right now is not a parsing function but a `dace.DeviceType` maker, and that's why the "Could not parse backend" message is weird. You could just write something like "'{backend}' is not a valid DaCe device type name"

**Review by egparedes — COMMENTED** [review]
I pressed the `send` button by mistake with only half of my comments. Here is the second half.

File: src/jace/util/jax_helper.py (hunk 15 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:38]: Before the update of the newest JAX we had to test for the array type, since for cupy we have to call `.toDLPack()` but for numpy it was necessary to call `.__dlpack__()`.
I simply forgot that.

File: src/jace/api.py (hunk 2 above)
Comment (philip-paul-mueller) [reply src/jace/api.py]: First a practical reason, if you read the backend string say from a file then you will get a `str`, so you will have to cast it.
It also does not really helps documenting since in _general_ it does not explain what a particular value mean or does, you still have to read the doc anyway.

File: src/jace/util/jax_helper.py [hunk 21]
```diff
@@ -258,3 +258,41 @@ def get_jax_literal_value(lit: jax_core.Atom) -> bool | float | int | np.generic
...
+        case "cpu" | "CPU":
+            return dace.DeviceType.CPU
+        case "gpu" | "GPU":
+            return dace.DeviceType.GPU
+        case "fpga" | "FPGA":
+            return dace.DeviceType.FPGA
+        case "tpu" | "TPU":
+            raise NotImplementedError("TPU are not supported.")
+        case _:
+            raise ValueError(f"Could not parse the backend '{backend}'.")
```
Comment (egparedes) [inline src/jace/util/jax_helper.py:279]: Rewrite the error message to reflect the new name of the function.

**Review by egparedes — APPROVED** [review]
LGTM. I have just a comment about an error message but I don't need to review it again so I'm already approving the PR now.

File: src/jace/util/jax_helper.py (hunk 21 above)
Comment (philip-paul-mueller) [reply src/jace/util/jax_helper.py:279]: You are absolutely right.
